Storage/Storage/SQL/BrowserDB.swift

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+public enum QuerySort {
+    case None, LastVisit, Frecency
+}
+
+public enum FilterType {
+    case None
+}
+
+public class QueryOptions {
+    // A filter string to apploy to the query
+    public var filter: String? = nil
+
+    // Allows for customizing how the filter is applied (i.e. only urls or urls and titles?)
+    public var filterType: FilterType = .None
+
+    // The way to sort the query
+    public var sort: QuerySort = .None
+
+    public init(filter: String? = nil, filterType: FilterType = .None, sort: QuerySort = .None) {
+        self.filter = filter
+        self.filterType = filterType
+        self.sort = sort
+    }
+}
+
+/* A table in our database. Note this doesn't have to be a real table. It might be backed by a join or something else interesting. */
+protocol Table {
+    typealias Type
+    var name: String { get }
+    func create(db: SQLiteDBConnection, version: Int) -> Bool
+    func updateTable(db: SQLiteDBConnection, from: Int, to: Int) -> Bool
+
+    func insert(db: SQLiteDBConnection, item: Type?, inout err: NSError?) -> Int
+    func update(db: SQLiteDBConnection, item: Type?, inout err: NSError?) -> Int
+    func delete(db: SQLiteDBConnection, item: Type?, inout err: NSError?) -> Int
+    func query(db: SQLiteDBConnection, options: QueryOptions?) -> Cursor
+}
+
+let DBCouldNotOpenErrorCode = 200
+
+/* This is a base interface into our browser db. It holds arrays of tables and handles basic creation/updating of them. */
+// Version 1 - Basic history table
+// Version 2 - Added a visits table, refactored the history table to be a GenericTable
+// Version 3 - Added a favicons table
```

Comment:
Add a comment to each table class that you need to come here, bump the version, and explain why if you make a change!